### PR TITLE
Add Timeout Feature to Email Notification Template

### PR DESCRIPTION
**Pull request by beeankha**
##### SUMMARY
Related to Issue https://github.com/ansible/awx/issues/2080

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME
 - API
 - UI

##### AWX VERSION

```
awx: 3.0.0
```

##### ADDITIONAL INFORMATION
Adding a timeout feature to email notifications using information found in Django base class found here: https://github.com/django/django/blob/stable/1.11.x/django/core/mail/backends/smtp.py

Adding a timeout scrollable input area in the UI.

### CODE REVIEW

**Review by jakemcdermott — DISMISSED** [review]
The code changes look good to me - Is this still WIP? 

Before merging you'll also want to give this a run with an email template that won't succeed just to confirm that providing the timeout in the notification config has the expected effect.

**Comment by beeankha** [comment]
Thank you for approving this, @jakemcdermott !   It is indeed still a WIP, testing the back end part of this to make sure that the timeout part of the notification is getting activated as needed.

**Comment by beeankha** [comment]
Below are the things I've tried so far to debug/figure out why failing notifications time out at 6 seconds no matter what.

**Logging**

Added logging to `email_backend.py`:

emailbackend_logger


**Remote Debugging**

Set a breakpoint with `sdb` in `email_backend.py`:
 
emailbackend_sdbtrace

Added an `sdb` breakpoint from inside of the AWX container to `smtp.py`: 

djangobackendsdbtrace


Nothing obvious is showing up yet, continuing to test this timeout feature...

File: awx/ui/client/src/notifications/notificationTemplates.form.js [hunk 1]
```diff
@@ -128,10 +128,25 @@ export default ['i18n', function(i18n) {
...
                 subForm: 'typeSubForm',
                 ngDisabled: '!(notification_template.summary_fields.user_capabilities.edit || canAdd)'
             },
+            timeout: {
+                label: i18n._('Timeout'),
+                type: 'number',
+                integer: true,
+                spinner: true,
+                min: 0,
+                default: 0,
```
Comment (kdelee) [inline awx/ui/client/src/notifications/notificationTemplates.form.js]: should the default timeout be 0? Seems like very little time

Comment (beeankha) [reply awx/ui/client/src/notifications/notificationTemplates.form.js]: It seems that a timeout value of 0 = "no timeout".  Still testing this, however.

**Comment by beeankha** [comment]
**Current Work**

- Testing to see what the min/max/default times should be
- Troubleshooting to see why there is an "in progress" fail message when the timeout is set to nothing (even when taking the `default: 0` out)
- Since this function is in there, does it now need to be required?  When left blank, it's failing every time.

**Progress That's Been Made**

There used to be only two types of notification status messages:

oldnotifcations


Now there are messages for Success, Timeout, and Fail (and the failure messages state WHY it didn't work):

newnotifications

**Comment by beeankha** [comment]
It appears that without the timeout field being an absolute requirement, this error gets thrown for what should be successful notifications:

operationnowinprogress

I tried to filter for this error by putting `res.data.error != "[Errno 115] Operation now in progress"` in the  `awx/ui/client/src/notifications/notification-templates-list/list.controller.js` file, here: https://github.com/ansible/awx/blob/bf1769af6c6eae01a1740258e63c27c3324990a2/awx/ui/client/src/notifications/notification-templates-list/list.controller.js#L155

...but that just triggered the `Call to test notifications failed.` error found here:
https://github.com/ansible/awx/blob/bf1769af6c6eae01a1740258e63c27c3324990a2/awx/ui/client/src/notifications/notification-templates-list/list.controller.js#L167

I tinkered with that line but I couldn't seem to bypass that error.  


The only thing I've been able to do to make the timeout error work is to make the field a requirement, and have it be a range of 5 - 120 seconds.  The max timeout can be longer, but anything shorter than 5 seconds throws the `[Errno 115] Operation now in progress` error.  

tooltip

**Comment by beeankha** [comment]
I made the timeout optional, but when there's no timeout stated, this error shows up:

CallToTestNotificationsFailed

This is still related to that `res.data.error != "[Errno 115] Operation now in progress"` error being filtered out in `awx/awx/ui/client/src/notifications/notification-templates-list/list.controller.js`:

OperationNowInProgress

**Comment by ryanpetrello** [comment]
One thing about this change (and the issue in general) that occurs to me is that I think a default of zero (no timeout) is actually a **really bad** default.

In my opinion, we should pick something sane, like 30 seconds, and let people change _that_ if they want.  I’d even argue **against** supporting “no timeout”, as it’s just going to result in hung processes if you've got your mail server misconfigured (bad).

```diff
diff --git a/awx/main/models/notifications.py b/awx/main/models/notifications.py
index 8bab75f0f..c0b4115be 100644
--- a/awx/main/models/notifications.py
+++ b/awx/main/models/notifications.py
@@ -187,7 +190,7 @@ class Notification(CreatedModifiedModel):
     )
     timeout = models.IntegerField(
         blank=True,
-        default=0,
+        default=30,
         help_text=_("The amount of time (in seconds) to run before the notification is canceled."),
     )
     body = JSONField(blank=True)
diff --git a/awx/ui/client/src/notifications/notificationTemplates.form.js b/awx/ui/client/src/notifications/notificationTemplates.form.js
index 3bc5b76d3..174d452cc 100644
--- a/awx/ui/client/src/notifications/notificationTemplates.form.js
+++ b/awx/ui/client/src/notifications/notificationTemplates.form.js
@@ -136,13 +136,13 @@ export default ['i18n', function(i18n) {
                 label: i18n._('Timeout'),
                 type: 'number',
                 integer: true,
-                min: 0,
-                default: 0,
+                min: 1,
+                default: 30,
                 spinner: true,
                 dataTitle: i18n._('Timeout'),
                 dataPlacement: 'right',
                 dataContainer: 'body',
-                awPopOver: "<p>" + i18n._("The amount of time (in seconds) before the email notification stops trying to reach the host and times out. Defaults to 0 for no time>
+                awPopOver: "<p>" + i18n._("The amount of time (in seconds) before the email notification stops trying to reach the host and times out.") + "</p>",
                 ngShow: "notification_type.value == 'email' ",
                 subForm: 'typeSubForm',
                 ngDisabled: '!(notification_template.summary_fields.user_capabilities.edit || canAdd)'
```

**Comment by beeankha** [comment]
As of right now, the `username` and `password` fields throw an error when they're blank, which is the next issue that needs to be fixed.  

usernamepasswordrequired

**Comment by beeankha** [comment]
Everything seems happy now.

EmailNotificationForm

Timeout default is 30 seconds, ranges from 1 - 120 seconds.

ToolTip

**Review by kialam — APPROVED** [review]
UI changes lgtm! Nice job.

File: awx/main/migrations/0064_notification_timeout.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
Comment (ryanpetrello) [inline awx/main/migrations/0064_notification_timeout.py]: On closer inspection, I don't think we actually _need_ this field (or migration) for the timeout to work.  We don't generally add backend-specific fields like this to the `main_notification` table unless all of the notifications use them (in the same way, you'll notice that `main_notification` doesn't have a username/password/host/port column).

The code that actually attempts to **send** the notification pulls the `timeout` (and other values) from the `main_notificationtemplate.notification_configuration` column:

```sql
awx-dev=> SELECT * FROM main_notificationtemplate;
-[ RECORD 1 ]--------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
id                         | 1
created                    | 2019-03-13 16:40:05.893517+00
modified                   | 2019-03-13 16:42:43.362269+00
description                |
name                       | Ryan Testing
notification_type          | email
notification_configuration | {"username": "user", "password": "$encrypted$UTF8$AESCBC$Z0FBQUFBQmNpVExLbFgyMm1VWEh2V1VOS3V6MTFQazgxYXU4a25hV21Pendud2RpNDk5VHN0bjBhUlFGZzQ3SHZRSWlCeEhkWHY5enVvQThtQnhIQS1LTXBGemdPT0NJU3c9PQ==", "host": "example.org", "recipients": ["joe@example.com"], "sender": "joe@example.com", "port": 25, "timeout": 3, "use_ssl": false, "use_tls": false}
created_by_id              | 1
modified_by_id             | 1
organization_id            | 1
```

https://github.com/ansible/awx/blob/devel/awx/main/models/notifications.py#L115-L128

With this diff:

```diff
diff --git a/awx/api/serializers.py b/awx/api/serializers.py
index 7dead78ea..ca454a0b1 100644
--- a/awx/api/serializers.py
+++ b/awx/api/serializers.py
@@ -4649,7 +4649,7 @@ class NotificationSerializer(BaseSerializer):
     class Meta:
         model = Notification
         fields = ('*', '-name', '-description', 'notification_template', 'error', 'status', 'notifications_sent',
-                  'notification_type', 'recipients', 'subject', 'timeout')
+                  'notification_type', 'recipients', 'subject')

     def get_related(self, obj):
         res = super(NotificationSerializer, self).get_related(obj)
diff --git a/awx/main/migrations/0064_notification_timeout.py b/awx/main/migrations/0064_notification_timeout.py
deleted file mode 100644
index 19f90eb04..000000000
--- a/awx/main/migrations/0064_notification_timeout.py
+++ /dev/null
@@ -1,20 +0,0 @@
-# -*- coding: utf-8 -*-
-# Generated by Django 1.11.16 on 2019-03-08 20:55
-from __future__ import unicode_literals
-
-from django.db import migrations, models
-
-
-class Migration(migrations.Migration):
-
-    dependencies = [
-        ('main', '0063_v350_org_host_limits'),
-    ]
-
-    operations = [
-        migrations.AddField(
-            model_name='notification',
-            name='timeout',
-            field=models.IntegerField(blank=True, default=0, help_text='The amount of time (in seconds) to run before the notification is canceled.'),
-        ),
-    ]
diff --git a/awx/main/migrations/0065_auto_20190312_1658.py b/awx/main/migrations/0065_auto_20190312_1658.py
deleted file mode 100644
index dd2edeae3..000000000
--- a/awx/main/migrations/0065_auto_20190312_1658.py
+++ /dev/null
@@ -1,20 +0,0 @@
-# -*- coding: utf-8 -*-
-# Generated by Django 1.11.16 on 2019-03-12 16:58
-from __future__ import unicode_literals
-
-from django.db import migrations, models
-
-
-class Migration(migrations.Migration):
-
-    dependencies = [
-        ('main', '0064_notification_timeout'),
-    ]
-
-    operations = [
-        migrations.AlterField(
-            model_name='notification',
-            name='timeout',
-            field=models.IntegerField(blank=True, default=30, help_text='The amount of time (in seconds) to run before the notification is canceled.'),
-        ),
-    ]
diff --git a/awx/main/models/notifications.py b/awx/main/models/notifications.py
index 6bd3e4c4e..39cea4917 100644
--- a/awx/main/models/notifications.py
+++ b/awx/main/models/notifications.py
@@ -185,11 +185,6 @@ class Notification(CreatedModifiedModel):
         default='',
         editable=False,
     )
-    timeout = models.IntegerField(
-        blank=True,
-        default=30,
-        help_text=_("The amount of time (in seconds) to run before the notification is canceled."),
-    )
     body = JSONField(blank=True)

     def get_absolute_url(self, request=None):
```

The timeout functionality seems to still work fine for me.

image

File: awx/main/models/notifications.py [hunk 3]
```diff
@@ -185,6 +185,11 @@ class Meta:
         default='',
         editable=False,
     )
+    timeout = models.IntegerField(
```
Comment (ryanpetrello) [inline awx/main/models/notifications.py]: Don't think we need this anymore, see big comment above.

File: awx/ui/client/src/notifications/notification-templates-list/list.controller.js [hunk 4]
```diff
@@ -131,7 +131,7 @@
                      } else {
                          ProcessErrors($scope, data, status, null, {
                              hdr: 'Error!',
-                             msg: 'Call to notifcatin templates failed. Notification returned status: ' + status
+                             msg: 'Call to notifcation templates failed. Notification returned status: ' + status
```
Comment (ryanpetrello) [inline awx/ui/client/src/notifications/notification-templates-list/list.controller.js]: Still has a typo 😄

File: awx/ui/client/src/notifications/notification-templates-list/list.controller.js [hunk 5]
```diff
@@ -152,9 +152,14 @@
...
                                  });
                                  $state.reload();
-                             } else if (res && res.data && res.data.status && res.data.status === "failed") {
+                             } else if (res && res.data && res.data.status && res.data.status === "failed" && res.data.error === "timed out") {
                                  ngToast.danger({
-                                     content: `<i class="fa fa-exclamation-triangle Toast-successIcon"></i> <b>${name}:</b> Notification failed.`
+                                     content: `<div><i class="fa fa-exclamation-triangle Toast-successIcon"></i> <b>${name}:</b> ${i18n._("Notification timed out.")}</div>`
+                                 });
+                                 $state.reload();
+                             } else if (res && res.data && res.data.status && res.data.status === "failed" && res.data.error !== "[Errno 115] Operation now in progress") {
```
Comment (ryanpetrello) [inline awx/ui/client/src/notifications/notification-templates-list/list.controller.js]: Don't think we need this block anymore.

File: awx/api/serializers.py [hunk 6]
```diff
@@ -4649,7 +4649,7 @@ class NotificationSerializer(BaseSerializer):
     class Meta:
         model = Notification
         fields = ('*', '-name', '-description', 'notification_template', 'error', 'status', 'notifications_sent',
-                  'notification_type', 'recipients', 'subject')
+                  'notification_type', 'recipients', 'subject', 'timeout')
```
Comment (ryanpetrello) [inline awx/api/serializers.py]: Probably don't need this, see comment below.

File: awx/ui/client/src/notifications/notification-templates-list/list.controller.js (hunk 5 above)
Comment (ryanpetrello) [reply awx/ui/client/src/notifications/notification-templates-list/list.controller.js]: Ah, sorry, I mean we don't need the `res.data.error !== "[Errno 115] Operation now in progress"` check.

File: awx/api/serializers.py (hunk 6 above)
Comment (beeankha) [reply awx/api/serializers.py]: 👍

File: awx/ui/client/src/notifications/notification-templates-list/list.controller.js (hunk 4 above)
Comment (beeankha) [reply awx/ui/client/src/notifications/notification-templates-list/list.controller.js]: 😂 
Fixing it!!

File: awx/main/models/notifications.py (hunk 3 above)
Comment (beeankha) [reply awx/main/models/notifications.py]: Removed and tested, everything was fine so this requested change has been made!

File: awx/ui/client/src/notifications/notification-templates-list/list.controller.js (hunk 4 above)
Comment (beeankha) [reply awx/ui/client/src/notifications/notification-templates-list/list.controller.js]: Fixed!

File: awx/ui/client/src/notifications/notification-templates-list/list.controller.js (hunk 5 above)
Comment (beeankha) [reply awx/ui/client/src/notifications/notification-templates-list/list.controller.js]: Removed as requested, tested it and everything still works fine so ✔️

File: awx/main/migrations/0064_notification_timeout.py (hunk 2 above)
Comment (beeankha) [reply awx/main/migrations/0064_notification_timeout.py]: I got the same results, so this migration file has been removed.

**Review by ryanpetrello — APPROVED** [review]

**Comment by ryanpetrello** [comment]
@tvo318 when we merge this, the email notifications creation UI will have a new field, `timeout`, which takes an integer.  It will default to 30s:

**Comment by elyezer** [comment]
I was able to check the new timeout option and it works as expected. One thing though is that if I set the timeout for more than 55~60 the following alert is shown:

Screenshot from 2019-03-14 14-37-04

So it seems that the call to test notification on the UI is timing our before the notification timeout and so the timeout alert:

Screenshot from 2019-03-14 14-38-48

Is never shown. With that said, I think the API is working fine but the UI still need to be updated.

**Comment by elyezer** [comment]
I tested with recent changes @beeankha sent and now the timeout notification is displaying as it should. With that said I think this PR is ready to be merged.

**Comment by beeankha** [comment]
recheck

**Comment by beeankha** [comment]
recheck

**Comment by ryanpetrello** [comment]
regate

**Pull request merged**